src/main/java/org/zalando/nakadi/validation/schema/SchemaDiff.java

```diff
@@ -0,0 +1,331 @@
+import static org.zalando.nakadi.domain.SchemaChange.Type.TITLE_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.TYPE_CHANGED;
+
+public class SchemaDiff {
+    public List<SchemaChange> collectChanges(final Schema original, final Schema update) {
+        final List<SchemaChange> changes = new ArrayList<>();
+        final Stack<String> jsonPath = new Stack<>();
+
+        recursiveCheck(original, update, jsonPath, changes);
+
+        return changes;
+    }
+
+    private void recursiveCheck(
+            final Schema original,
+            final Schema update,
+            final Stack<String> jsonPath,
+            final List<SchemaChange> changes) {
+
+        if (original == null && update == null) {
+            return;
+        }
+
+        if (original == null || update == null) {
+            addChange(SCHEMA_REMOVED, jsonPath, changes);
+            return;
+        }
+
+        if (!original.getClass().equals(update.getClass())) {
+            addChange(TYPE_CHANGED, jsonPath, changes);
+            return;
+        }
+
+        if (!Objects.equals(original.getId(), update.getId())) {
+            addChange(ID_CHANGED, jsonPath, changes);
+        }
+
+        if (!Objects.equals(original.getTitle(), update.getTitle())) {
+            addChange(TITLE_CHANGED, jsonPath, changes);
+        }
+
+        if (!Objects.equals(original.getDescription(), update.getDescription())) {
+            addChange(DESCRIPTION_CHANGED, jsonPath, changes);
+        }
+
+        if (original instanceof StringSchema) {
+            recursiveCheck((StringSchema) original, (StringSchema) update, jsonPath, changes);
+        } else if (original instanceof NumberSchema) {
+            recursiveCheck((NumberSchema) original, (NumberSchema) update, jsonPath, changes);
+        } else if (original instanceof EnumSchema) {
+            recursiveCheck((EnumSchema) original, (EnumSchema) update, jsonPath, changes);
+        } else if (original instanceof CombinedSchema) {
+            recursiveCheck((CombinedSchema) original, (CombinedSchema) update, jsonPath, changes);
+        } else if (original instanceof ObjectSchema) {
+            recursiveCheck((ObjectSchema) original, (ObjectSchema) update, jsonPath, changes);
+        } else if (original instanceof ArraySchema) {
+            recursiveCheck((ArraySchema) original, (ArraySchema) update, jsonPath, changes);
+        } else if (original instanceof ReferenceSchema) {
+            recursiveCheck((ReferenceSchema) original, (ReferenceSchema) update, jsonPath, changes);
+        }
+    }
+
+    private void recursiveCheck(final StringSchema stringSchemaOriginal, final StringSchema stringSchemaUpdate,
+                                final Stack<String> jsonPath, final List<SchemaChange> changes) {
+        if (!Objects.equals(stringSchemaOriginal.getMaxLength(), stringSchemaUpdate.getMaxLength())) {
+            jsonPath.push("maxLength");
```

Comment:
`jsonPath` is a `Stack` object where I push the attribute under inspection. Since it has recursion, once it comes out of recursion, I pop the attribute.

It is used to keep a memory of the path of the recursion for reporting the where in the JSON the invalid attribute was found.

I could create an auxiliary method that would receive a `keyword` as string and a lambda, would push the `keyword` execute the lambda and pop the `keyword`. That could reduce the chances of forgetting to pop something.